### ISSUE TITLE
Disallow unsafe cast() uses.

### ISSUE THREAD

**Issue by jenstroeger** [python/mypy#14797]
### Feature

Certain uses of the [`cast()`](https://docs.python.org/3/library/typing.html#typing.cast) function should be disallowed and rejected by the type checker. In fact, I think that “unsafe” uses of `cast()` should be an explicit _opt-in_ feature.

This feature request seems related to https://github.com/python/mypy/issues/5756

### Pitch

Today I came across the following statement (yes, this is production code):
```python
foo: Foo = cast(Foo, None)
```
embedded in a function roughly like so:
```python
def get_foo() -> Foo:
    for foo in some_collection:
        if condition:
            return foo
    return cast(Foo, None)
```
Allowing to cast `None` to another type is the essence of an “unsafe” cast and should not be allowed. Furthermore, I’d consider the following uses of the `cast()` function nonsense as well:
```python
s: str = typing.cast(str, 5)
reveal_type(s)

n: None = typing.cast(None, "foo")
reveal_type(n)

i: int = typing.cast(int, 3.1415)
reveal_type(i)

class Foo:
    pass

f: Foo = typing.cast(Foo, None)
reveal_type(f)

class Bar:
    pass

b: Bar = typing.cast(Bar, Foo())
reveal_type(b)
```
which actually gives me
```
note: Revealed type is "builtins.str"
note: Revealed type is "None"
note: Revealed type is "builtins.int"
note: Revealed type is "test.Foo"
note: Revealed type is "test.Bar"
Success: no issues found in 1 source file
```
If somebody — for whatever reason — insists on such an unsafe type cast, then they should make their intentions clear, e.g.
```python
foo: Foo = cast(Foo, None)  # type: ignore[unsafe-cast]
```
Allowing the above casts, to me, defeats the purpose of using a type checker as a code QA tool which is supposed to help prevent bugs.

**Comment by hauntsaninja** [comment]
I broadly agree.

There are two uses of safe casts:
- The first is an upcast, which allows the user to safely erase a narrowed type https://github.com/python/mypy/issues/5756
- The second is just providing a type context to help mypy's inference (similar to how PEP 526 annotations provide type context)

Everything else is unsafe.

If you think of types as sets, I can see that it could be useful to disambiguate casts to an overlapping type (where there is a value that can possibly belong to the casted type), such as downcasts, NewType, cast to TypedDict, etc — from your "nonsense" casts where the sets of values are completely disjoint.

I would potentially be okay with including an opt-in check to detect nonsense casts in mypy.

But I do kind of want to pursue adding differently named functions for these three cases, or at least for the safe cast vs unsafe cast. In particular, I find users are sometimes completely unaware of how unsafe `cast` is; I think the name can be somewhat innocuous for users who haven't worked with C / C++.

**Comment by jenstroeger** [comment]
> But I do kind of want to pursue adding differently named functions for these three cases, or at least for the safe cast vs unsafe cast.

I agree with adding a `safe_cast()` and maybe `unsafe_cast()` function and have `mypy` raise issue when `safe_cast()` is used for unsafe casts. And I guess the existing `cast()` function should then be considered “unsafe” (although it does enable the nonsense category of casts 😉)?

And we’re circling in on discussion #5756

> In particular, I find users are sometimes completely unaware of how unsafe `cast` is; I think the name can be somewhat innocuous for users who haven't worked with C / C++.

Hah, yeah. The initial `cast(Foo, None)` came from somebody whose background is C/C++

**Comment by erictraut** [comment]
The semantics and behavior for the existing `cast` are well established, so I don't think it would be a good idea to redefine the behavior, even as an optional setting. I agree that `cast` is very dangerous, and I frequently encourage Python developers to use it only as a last resort.

Adding a safer form of cast is a reasonable idea, but it would be better to move that discussion to the [python/typing discussion forum](https://github.com/python/typing/discussions) because such an addition would presumably need to be added to the python type system and would not be specific to mypy.

**Comment by hauntsaninja** [comment]
Yeah, let's move discussion to https://github.com/python/typing/issues/565

**Comment by jenstroeger** [comment]
> @erictraut Adding a safer form of cast is a reasonable idea, but it would be better to move that discussion to the [python/typing discussion forum](https://github.com/python/typing/discussions) because such an addition would presumably need to be added to the python type system and would not be specific to mypy.

Fair enough. Should I open a discussion with the above blurb there again, or what’s the recommended approach?

> @hauntsaninja Yeah, let's move discussion to https://github.com/python/typing/issues/565

Thanks for the reference, that issue is interesting and quite related; I might chime in for a moment. So we now have https://github.com/python/mypy/issues/5756 and https://github.com/python/typing/issues/565 and the [python/typing discussion forum](https://github.com/python/typing/discussions) to continue this conversation 🤓

An alternative would be a writing a [flake8 plugin](https://flake8.pycqa.org/en/latest/plugin-development/) or a [pylint plugin](https://pylint.pycqa.org/en/latest/development_guide/how_tos/plugins.html) instead… at least for the aforementioned uses of `cast()` involving builtin types.
